tempesta_fw/ptrset.h

```diff
@@ -0,0 +1,296 @@
+			 __tfw_ptrset_checkptr(s, ptr), \
+			 __tfw_ptrset_max(s))
+
+#define tfw_ptrset_del(s, ptr) \
+	__tfw_ptrset_del(__tfw_ptrset_cast(s), \
+			 __tfw_ptrset_checkptr(s, ptr), \
+			 __tfw_ptrset_max(s))
+
+#define tfw_ptrset_test(s, ptr) \
+	(0 < __tfw_ptrset_find_idx(__tfw_ptrset_case(s), \
+				   __tfw_ptrset_checkptr(s, ptr), \
+				   __tfw_ptrest_max(s)))
+
+/**
+ * Sanity/debug runtime checks.
+ * The function is completely eliminated by the compiler when DEBUG is not
+ * defined, so it gives no overhead.
+ */
+static inline void
+__tfw_ptrset_validate(const TfwPtrSet *s, int max)
+{
+	IF_DEBUG {
+		void *ptr1, *ptr2;
+		int idx1, idx2;
+
+		BUG_ON(!s);
+		BUG_ON(max > 4096); /* Large value is likely garbage. */
+		BUG_ON(max <= 0);   /* Can't be zero unlike TfwPtrSet->max. */
+		BUG_ON(s->n < 0 || s->n > max);
+		BUG_ON(s->max < 0 || s->max > 4096);
+
+		tfw_ptrset_for_each(ptr1, idx1, s) {
+			BUG_ON(!ptr1);
+			tfw_ptrset_for_each(ptr2, idx2, s) {
+				BUG_ON(ptr1 == ptr2 && idx1 != idx2);
+			}
+		}
+	}
+}
+
+static inline int
+__tfw_ptrset_find_idx(const TfwPtrSet *s, const void *ptr, int max)
+{
+	int i;
+
+	__tfw_ptrset_validate(s, max);
+
+	for (i = 0; i < max; ++i) {
+		if (s->ptrs[i] == ptr)
+			return i;
+	}
+
+	return -1;
+}
+
+static inline void
+__tfw_ptrset_init(TfwPtrSet *s, int max)
+{
+	memset(s->ptrs, 0, tfw_ptrset_size(max));
+	s->max = max;
+	__tfw_ptrset_validate(s, max);
+}
+
+static inline int
+__tfw_ptrset_add(TfwPtrSet *s, void *ptr, int max)
+{
+	if (__tfw_ptrset_find_idx(s, ptr, max) > 0) {
+		TFW_ERR("Can't add ptr %p to set %p - duplicate ptr\n", ptr, s);
+		return -1;
+	}
+	else if (s->n >= s->max) {
+		TFW_ERR("Can't add ptr %p to set %p - set is full\n", ptr, s);
+		return -1;
+	}
+
+	s->ptrs[s->n] = ptr;
+	++s->n;
+
+	return 0;
+}
+
+static inline int
+__tfw_ptrset_del(TfwPtrSet *s, const void *ptr, int max)
+{
+	int i;
+
+	i = __tfw_ptrset_find_idx(s, ptr, max);
+	if (i < 0) {
+		TFW_ERR("Can't delete ptr %p from set %p - not found\n", ptr,
+			s);
+		return -1;
+	}
+
+	s->ptrs[i] = s->ptrs[s->n - 1];
+	s->ptrs[s->n] = NULL;
+	--s->n;
+
+	return 0;
+}
+
+/**
+ * Return pointers from the @s in the round-robin manner.
+ *
+ * On each call you get the next pointer in the set, and so on in cycle.
+ *
+ * The order of pointers is not guaranteed since tfw_ptrset_del() changes it.
+ * Not very fair, but still good unless you delete elements often.
+ *
+ * No synchronization with add/delete/purge operations is needed (lock-free).
+ *
+ * Safe for use on multiple CPUs because of the atomic operations.
+ * BUG: atomic_t contains signed int which is overflown.
+ *      Perhaps we should use a regular unsigned int instead of atomic_t here?
+ *      That fixes the bug, and also makes the code faster, but less fair.
+ */
+#define tfw_ptrset_get_rr(s) \
+	__tfw_ptrset_get_rr(__tfw_ptrset_cast(s), __tfw_ptrset_max(s))
+
+static inline void *
+__tfw_ptrset_get_rr(TfwPtrSet *s, int max)
```

Comment:
PtrSet is very specific. The whole it's sens in this function and it's very specific for that request scheduler. So it seems it has no sense to move it to separate library-like file.